File: android/revenuedog/src/main/kotlin/com/revenuedog/purchases/models/Offering.kt
package com.revenuedog.purchases.models

/**
 * A paywall configuration: an ordered list of [Package]s the app can present.
 *
 * Packages whose product Google Play could not return are dropped before the offering
 * is surfaced — a package always has a price.
 */
class Offering internal constructor(
    val identifier: String,
    val serverDescription: String,
    /** Free-form metadata configured on the backend. */
    val metadata: Map<String, Any>,
    val availablePackages: List<Package>
) {
    /** The `$rd_lifetime` package, if present. */
    val lifetime: Package? get() = byType(PackageType.LIFETIME)

    /** The `$rd_annual` package, if present. */
    val annual: Package? get() = byType(PackageType.ANNUAL)

    /** The `$rd_six_month` package, if present. */
    val sixMonth: Package? get() = byType(PackageType.SIX_MONTH)

    /** The `$rd_three_month` package, if present. */
    val threeMonth: Package? get() = byType(PackageType.THREE_MONTH)

    /** The `$rd_two_month` package, if present. */
    val twoMonth: Package? get() = byType(PackageType.TWO_MONTH)

    /** The `$rd_monthly` package, if present. */
    val monthly: Package? get() = byType(PackageType.MONTHLY)

    /** The `$rd_weekly` package, if present. */
    val weekly: Package? get() = byType(PackageType.WEEKLY)

    /** Looks up a package by its full identifier (reserved or custom). */
    fun getPackage(identifier: String): Package? =
        availablePackages.firstOrNull { it.identifier == identifier }

    private fun byType(type: PackageType): Package? =
        availablePackages.firstOrNull { it.packageType == type }

    override fun toString(): String =
        "Offering(identifier=$identifier, packages=${availablePackages.map { it.identifier }})"
}

File: ios/Sources/RevenueDog/Models/Offering.swift
import Foundation

/// A named collection of purchasable ``Package``s, typically backing one
/// paywall.
public struct Offering {
    /// The offering identifier (e.g. `"default"`).
    public let identifier: String
    /// The server-provided description for this offering.
    public let serverDescription: String
    /// Arbitrary server-defined metadata.
    public let metadata: [String: Any]
    /// The packages available in this offering. Packages whose product the
    /// store could not resolve are omitted.
    public let availablePackages: [Package]

    public init(identifier: String,
                serverDescription: String,
                metadata: [String: Any],
                availablePackages: [Package]) {
        self.identifier = identifier
        self.serverDescription = serverDescription
        self.metadata = metadata
        self.availablePackages = availablePackages
    }

    /// Look up a package by its RevenueDog identifier.
    public func package(identifier: String) -> Package? {
        availablePackages.first { $0.identifier == identifier }
    }

    private func package(ofType type: PackageType) -> Package? {
        availablePackages.first { $0.packageType == type }
    }

    /// The `$rd_lifetime` package, if present.
    public var lifetime: Package?   { package(ofType: .lifetime) }
    /// The `$rd_annual` package, if present.
    public var annual: Package?     { package(ofType: .annual) }
    /// The `$rd_six_month` package, if present.
    public var sixMonth: Package?   { package(ofType: .sixMonth) }
    /// The `$rd_three_month` package, if present.
    public var threeMonth: Package? { package(ofType: .threeMonth) }
    /// The `$rd_two_month` package, if present.
    public var twoMonth: Package?   { package(ofType: .twoMonth) }
    /// The `$rd_monthly` package, if present.
    public var monthly: Package?    { package(ofType: .monthly) }
    /// The `$rd_weekly` package, if present.
    public var weekly: Package?     { package(ofType: .weekly) }
}

// Allow value comparison in tests / consumers (metadata is excluded as
// `[String: Any]` is not Equatable).
extension Offering: Equatable {
    public static func == (lhs: Offering, rhs: Offering) -> Bool {
        lhs.identifier == rhs.identifier &&
        lhs.serverDescription == rhs.serverDescription &&
        lhs.availablePackages.map(\.identifier) == rhs.availablePackages.map(\.identifier)
    }
}
